File: src/main/scala/uk/co/ridentbyte/view/request/RequestUriVerbInputPane.scala
package uk.co.ridentbyte.view.request

import javafx.collections.FXCollections
import javafx.scene.control.{ChoiceBox, TextField}
import javafx.scene.layout.{GridPane, Priority}

import scala.collection.JavaConverters._

class RequestUriVerbInputPane(triggerUnsavedChangesMade: () => Unit) extends GridPane {

  setHgap(10)
  setVgap(10)

  private val textUri = new TextField()
  textUri.getStyleClass.add("cardinal-font")
  textUri.textProperty().addListener((_, _, _) => triggerUnsavedChangesMade())
  GridPane.setVgrow(textUri, Priority.NEVER)
  GridPane.setHgrow(textUri, Priority.ALWAYS)
  textUri.setPromptText("http://localhost:8080")
  add(textUri, 0, 0)

  private val selectVerb = new ChoiceBox[String](FXCollections.observableArrayList("GET", "POST", "PUT", "DELETE", "HEAD", "CONNECT", "OPTIONS", "TRACE", "PATCH"))
  selectVerb.getStyleClass.add("cardinal-font")
  selectVerb.getSelectionModel.selectFirst()
  selectVerb.getSelectionModel.selectedItemProperty.addListener((_, _, _) => triggerUnsavedChangesMade())
  GridPane.setVgrow(selectVerb, Priority.NEVER)
  GridPane.setHgrow(selectVerb, Priority.NEVER)
  add(selectVerb, 1, 0)

  def getUri: String =  textUri.getText.trim

  def getVerb: String = selectVerb.getSelectionModel.getSelectedItem

  def setUri(uri: String): Unit = textUri.setText(uri)

  def setVerb(verb: String): Unit = {
    val matchingIndex = selectVerb.getItems.asScala.zipWithIndex.find {
      case (item: String, _) => item == verb
    }.map { verbWithIndex =>
      verbWithIndex._2
    }.getOrElse(0)
    selectVerb.getSelectionModel.select(matchingIndex)
  }

  def clear(): Unit = {
    textUri.setText("")
    selectVerb.getSelectionModel.select(0)
  }

}

File: src/main/java/uk/co/ridentbyte/view/request/RequestUriVerbInputPane.java
package uk.co.ridentbyte.view.request;

import javafx.collections.FXCollections;
import javafx.scene.control.CheckBox;
import javafx.scene.control.ChoiceBox;
import javafx.scene.control.TextField;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import uk.co.ridentbyte.functions.UnsavedChangesMade;

public class RequestUriVerbInputPane extends GridPane {

    private TextField textUri;
    private ChoiceBox<String> selectVerb;
    private CheckBox followRedirects;

    public RequestUriVerbInputPane(UnsavedChangesMade unsavedChangesMade) {
        this.setHgap(10);
        this.setVgap(10);

        this.textUri = new TextField();
        this.textUri.getStyleClass().add("cardinal-font");
        this.textUri.textProperty().addListener((arg, oldVal, newVal) -> unsavedChangesMade.trigger());
        this.textUri.setPromptText("http://localhost:8080");
        GridPane.setVgrow(this.textUri, Priority.NEVER);
        GridPane.setHgrow(this.textUri, Priority.ALWAYS);
        this.add(this.textUri, 0, 0);

        this.selectVerb = new ChoiceBox<>(FXCollections.observableArrayList("GET", "POST", "PUT", "DELETE", "HEAD", "OPTIONS", "TRACE", "PATCH"));
        this.selectVerb.getStyleClass().add("cardinal-font");
        this.selectVerb.getSelectionModel().selectFirst();
        this.selectVerb.getSelectionModel().selectedItemProperty().addListener((arg, oldVal, newVal) -> unsavedChangesMade.trigger());
        GridPane.setVgrow(this.selectVerb, Priority.NEVER);
        GridPane.setHgrow(this.selectVerb, Priority.NEVER);
        add(this.selectVerb, 1, 0);

        this.followRedirects = new CheckBox("Follow redirects");
        this.followRedirects.selectedProperty().addListener((arg) -> unsavedChangesMade.trigger());
        add(this.followRedirects, 0, 1);
    }

    public String getUri() {
        return this.textUri.getText().trim();
    }

    public void setUri(String uri) {
        this.textUri.setText(uri);
    }

    public String getVerb() {
        return this.selectVerb.getSelectionModel().getSelectedItem();
    }

    public boolean shouldFollowRedirects() {
        return this.followRedirects.isSelected();
    }

    public void setVerb(String verb) {
        int index = this.selectVerb.getItems().indexOf(verb);
        if (index >= 0) {
            this.selectVerb.getSelectionModel().select(index);
        }
    }

    public void clear() {
        this.textUri.setText("");
        this.selectVerb.getSelectionModel().select(0);
    }

    public void setShouldFollowRedirects(boolean shouldFollowRedirects) {
        this.followRedirects.setSelected(shouldFollowRedirects);
    }
}
